src/Pact/Utils/Json.hs

```diff
@@ -0,0 +1,211 @@
+--
+legacyHashMap
+  :: forall a b
+  . HasCallStack
+  => Typeable a
+  => ToJSONKey a
+  => HM.HashMap a b
+  -> HM.HashMap (LegacyHashed T.Text) b
+legacyHashMap = case toJSONKey @a of
+  ToJSONKeyText f _ -> HM.mapKeys (LegacyHashed . f)
+  _ -> error $ "Pact.Utils.Json: failed to JSON encode HashMap with non-textual key of type " <> show (typeRep (Proxy @a))
+
+-- | Convert a 'M.Map' that has a textual JSON key into HashMap with a Legacy
+-- Key.
+--
+-- The 'ToJSON' instance for Map encodes maps with textual keys as 'Object',
+-- which internally is a 'HM.HashMap' with 'T.Text' as keys. For 'toEncoding' to
+-- preserve the legacy ordering of properties with convert the map into
+-- 'HM.HashMap' with 'LegacyHashed T.Text' as key type.
+--
+-- Apply this function to 'M.Map' values in 'JsonProperties' functions.
+--
+legacyMap
+  :: forall a b
+  . HasCallStack
+  => Typeable a
+  => ToJSONKey a
+  => M.Map a b
+  -> HM.HashMap (LegacyHashed T.Text) b
+legacyMap = case toJSONKey @a of
+  ToJSONKeyText f _ -> HM.fromList . fmap (first (LegacyHashed . f)) . M.toList
+  _ -> error $ "Pact.Utils.Json: failed to JSON encode Map with non-textual key of type " <> show (typeRep (Proxy @a))
+
+-- -------------------------------------------------------------------------- --
+-- Hashing of 'T.Text' for `text <2` and `hashable <1.3.1`
+
+-- | Legacy hash function for 'T.Text', that preserves the semantics of
+-- text-1.5 and hashable-1.3.0 on 64bit POSIX systems.
+--
+legacyHash :: T.Text -> Int
+legacyHash = fromIntegral . hashable_fnv1 . T.encodeUtf16LE
+{-# INLINE legacyHash #-}
+
+legacyHashWithSalt :: Int -> T.Text -> Int
+legacyHashWithSalt salt =
+  fromIntegral . hashable_fnv1_withSalt (fromIntegral salt) . T.encodeUtf16LE
+{-# INLINE legacyHashWithSalt #-}
+
+-- | Non-optimized implementation of FNV1 as used in hashable <1.3.1.
+-- Only use for short inputs!
+--
+-- This implemetation deviates from the standard as follows
+--
+-- * It uses the 32 prime with 64bit arithmetic (on architectures where long has 8 bytes)
+-- * It initializes the hash with a custom salt instead of FNV_offset_basis
+--
+hashable_fnv1_withSalt :: Int64 -> B.ByteString -> Int64
+hashable_fnv1_withSalt = B.foldl' go
+ where
+  go :: Int64 -> Word8 -> Int64
+  go h c = (h * fnv_prime_32) `xor` fromIntegral c
+
+  -- 32 bit fnv1 prime
+  fnv_prime_32 :: Int64
+  fnv_prime_32 = 16777619
+
+hashable_fnv1 :: B.ByteString -> Int64
+hashable_fnv1 = hashable_fnv1_withSalt hashable_default_salt
+ where
+  hashable_default_salt :: Int64
+  hashable_default_salt = -2578643520546668380
+{-# INLINE hashable_fnv1 #-}
+
+-- -------------------------------------------------------------------------- --
+--
+
+newtype LegacyValue = LegacyValue { getLegacyValue :: Value }
+  deriving (Show, Eq, NFData)
+
+instance ToJSON LegacyValue where
+  toJSON v = error $ "Pact.Utils.Json.LegacyValue: attempt to call toJSON on " <> show v
```

Comment:
It's hard to tell just from the code -- maybe somebody who is more familiar with the entire pact code base (ping @jmcardon) could take a look. 

Additionally, the PR  also includes quick check test cases for all `ToJSON` instances in the pact code base for checking that we never call `toJSON` on `LegacyValue`.

Reply:
I've seen this being triggered in chainweb tests. I assume it's only in testing, but I haven't investigated this more in depth yet. It should be possible to avoid it in prod.